pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -413,27 +415,124 @@ private PinotResourceManagerResponse uploadSegment(File indexDir, SegmentMetadat
+
+    long minionExpectedCrc = -1L;
+    boolean isMinionConfigured = offlineTableConfig.getTaskConfig() != null;
+    boolean minionHeaderPresent = headers.getHeaderString(HttpHeaders.USER_AGENT).contains(CommonConstants.Minion.MINION_HEADER);
+    try {
+      if (isMinionConfigured) {
+        // Minion configured use case so that we write to zk sparingly
+        String segmentPushStatus = originalOfflineSegmentZkMetadata.getSegmentPushStatus();
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd'T'HH:mm:ss.SSS");
+        Date dt = new Date();
+        String currentHumanReadableTime = sdf.format(dt);
+        offlineSegmentZKMetadataWithPushStatus.setSegmentPushStatus("COMMITTING-" + currentHumanReadableTime);
+
+        if (segmentPushStatus != null) {
+          if (tooLate(dt)) {
+            LOGGER.error("Segment was not properly committed, replacing segment {} for table {}", segmentName,
+                tableName);
+            _controllerMetrics.addMeteredGlobalValue(ControllerMeter.CONTROLLER_SEGMENT_UPLOAD_ERROR, 1L);
+          } else {
+            throw new ControllerApplicationException(LOGGER, "Please retry later, segment upload in progress", Response.Status.INTERNAL_SERVER_ERROR);
+          }
+        }
+
+        if (!_pinotHelixResourceManager
+            .updateZkMetadataAtomically(offlineSegmentZKMetadataWithPushStatus, znRecord.getVersion())) {
+          // Write committing status to zk
+          throw new ControllerApplicationException(LOGGER, "Retry later. ZK update failed.", Response.Status.INTERNAL_SERVER_ERROR);
+        }
+
+        try {
+          minionExpectedCrc = Long.parseLong(headers.getHeaderString(HttpHeaders.IF_MATCH));
+        } catch (NumberFormatException e) {
+          LOGGER.warn("Could not parse crc {}", minionExpectedCrc, e);
+          throw new ControllerApplicationException(LOGGER, "Invalid crc sent from minion", Response.Status.NOT_ACCEPTABLE);
+        }
+      }
+
+      if (downloadUrl == null) {
+        downloadUrl = moveSegmentToPermanentDirectory(provider, tableName, segmentName, tempTarredSegmentFile);
+      }
+
+      if (shouldRefreshSegment(segmentMetadata, originalOfflineSegmentZkMetadata, offlineTableName, segmentName,
+          minionExpectedCrc, minionHeaderPresent)) {
+        response = _pinotHelixResourceManager.refreshSegment(segmentMetadata, originalOfflineSegmentZkMetadata, downloadUrl);
+      } else {
+        response = PinotResourceManagerResponse.SUCCESS_RESPONSE;
+      }
+
+      if (!response.isSuccessful()) {
+        _controllerMetrics.addMeteredGlobalValue(ControllerMeter.CONTROLLER_SEGMENT_UPLOAD_ERROR, 1L);
+        throw new ControllerApplicationException(LOGGER, "Error uploading segment", Response.Status.INTERNAL_SERVER_ERROR);
+      }
+      return response;
+    } catch (ControllerApplicationException e) {
+      throw e;
+    } finally {
+      if (isMinionConfigured) {
```

Comment:
You need to check against whether or not the metadata was updated. Not if (minionConfigured)